### PR TITLE
pubsub: add ErrorAs

### LINKED ISSUE #957: pubsub: add ErrorAs()

### PR DESCRIPTION
**Pull request by jba**
- Add an ErrorAs method to Topic and Subscription.
- Add it to drivers as well.
- Write test.

Fixes #957.

### CODE REVIEW

File: pubsub/drivertest/drivertest.go [hunk 1]
```diff
@@ -94,6 +99,10 @@ func (verifyAsFailsOnNil) SubscriptionCheck(s *pubsub.Subscription) error {
 	return nil
 }
 
+func (verifyAsFailsOnNil) ErrorCheck(t *pubsub.Topic, err error) error {
+	return nil
```
Comment (vangent) [inline pubsub/drivertest/drivertest.go]: This should call `t.ErrorAs(nil)` and verify it returns false (like the other tests in the struct.

File: pubsub/pubsub.go [hunk 2]
```diff
@@ -166,6 +166,27 @@ func (t *Topic) As(i interface{}) bool {
 	return t.driver.As(i)
 }
 
+// ErrorAs converts err to provider-specific types.
+// See Topic.As for more details.
+// ErrorAs panics if target is nil or not a pointer.
```
Comment (vangent) [inline pubsub/pubsub.go]: Be consistent with `blob`... I don't think panics in those scenarios does it?

File: pubsub/rabbitpubsub/rabbit.go [hunk 3]
```diff
@@ -340,6 +346,21 @@ func (t *topic) As(i interface{}) bool {
 	return true
 }
 
+// ErrorAs implements driver.Topic.ErrorAs
+func (*topic) ErrorAs(err error, target interface{}) bool {
+	return errorAs(err, target)
+}
+
+func errorAs(err error, target interface{}) bool {
+	if aerr, ok := err.(*amqp.Error); ok {
```
Comment (vangent) [inline pubsub/rabbitpubsub/rabbit.go]: Support `MultiError` here as well?

File: pubsub/pubsub.go (hunk 2 above)
Comment (jba) [reply pubsub/pubsub.go]: Could I change blob? Both should be consistent with `xerrors.As`. See https://godoc.org/golang.org/x/exp/xerrors#As. Note that there is now an [official proposal](https://github.com/golang/go/issues/29934) for this.

Comment (vangent) [reply pubsub/pubsub.go]: Sure. And `runtimevar`.

File: pubsub/rabbitpubsub/rabbit.go (hunk 3 above)
Comment (jba) [reply pubsub/rabbitpubsub/rabbit.go]: Done.

File: pubsub/gcppubsub/gcppubsub.go [hunk 4]
```diff
@@ -137,6 +139,21 @@ func (t *topic) As(i interface{}) bool {
 	return true
 }
 
+// ErrorAs implements driver.Topic.ErrorAs
+func (*topic) ErrorAs(err error, target interface{}) bool {
+	return errorAs(err, target)
+}
+
+func errorAs(err error, target interface{}) bool {
+	if s, ok := status.FromError(err); ok {
```
[the remark this answers is not in the document]
Comment (jba) [reply pubsub/gcppubsub/gcppubsub.go]: Done.

File: pubsub/rabbitpubsub/rabbit.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jba) [reply pubsub/rabbitpubsub/rabbit.go]: Rewritten with a switch.

File: pubsub/drivertest/drivertest.go (hunk 1 above)
Comment (jba) [reply pubsub/drivertest/drivertest.go]: Replaced with a panic check.

**Comment by jba** [comment]
PTAL.

File: pubsub/rabbitpubsub/doc.go [hunk 5]
```diff
@@ -28,4 +28,5 @@
 // As
 //  - Topic: *amqp.Connection
 //  - Subscription: *amqp.Connection
+//  - Error: *amqp.Error
```
Comment (vangent) [inline pubsub/rabbitpubsub/doc.go]: and ampq.MultiError.

**Comment by jba** [comment]
PTAL @ijt

**Comment by jba** [comment]
@ijt ping

**Pull request merged**